content/blog/2019-10-23-loop-unrolling-with-bril.md

```diff
@@ -0,0 +1,154 @@
+
+To obtain the initial value of the induction variable, we check the result of constant propagation in the *entry* block. 
+Notice that we only care about the **out** sets of basic blocks that are (1) not in the loop, and (2) can reach the *entry* block. If we can obtain one single constant value of the induction variable by examining all these out sets, then we have 
+successfully determined the initial value of the induction variable. Otherwise, the loop won't be unrolled. 
+
+Now the tripcount can be computed by ``(bound - initial_value + stride - 1) // stride``. Notice that with this formula, a 
+negative tripcount refers to an infinite loop. 
+
+##### Unrolling the Loop
+We can finally unroll the loop now. Since we have inserted a label for each basic block, we can simply duplicate the basic
+blocks inside the loop and redirect the control flow when it reaches the backedge. If the control flow can exit the loop from the *entry*, we need to duplicate the *entry* block one more time. All but the last duplicates of the ``br`` instruction for the back-edge can be safely converted to ``jmp`` instructions. 
+
+The generated code would be functionally correct. 
+However, it is not very efficient because of the redundant ``br`` and ``jmp`` instructions: we already know where the control flow will go to at the backedge. Therefore, it is ideal to merge the basic blocks of the unrolled loop as much as 
+possible. We use a trick to achieve this: rather than actually merging the basic blocks in the code, we leverage the orderedness of the ``OrderedDict`` and ``list`` data structures in Python to force a meaningful order of basic block 
+placement. What we actually do is to remove the labels of all but the first duplicate of the *entry* block and all the *exit* blocks, and remove all the ``jmp`` instructions of all the intermediate duplicates of the *entry* and *exit* blocks. 
+
+#### Dynamic Profiling 
+
+The dynamic profiling based unrolling pass consists of two parts mainly: loop status analysis, program structure re-construction. In the loop status analysis part, we created a variable table to track the defition and usage for each varible. The table is used as a look-up table for loop analysis and basic block replication in later stages. The variable table can be extended to record the data dependency between different varibles, and this information can be used to realize loop normalization, copy propagation and such. With the update information for each variables, we then iterate all the back edges in the control flow graph and unroll the loops accordingly. Note that the back edges are stored in a dictionary and need to be processed in order to make sure the loops can be unrolled correctly (e.g. for nested loops, the inner most loop needs to be unrolled first). In the last step, we replicate the loop body and remove the condition branch instructions as well as the back edges to construct the new unrolled program.
```

Comment:
I'm a little confused about this paragraph—it's under a heading called "Dynamic Profiling," but it doesn't actually say anything about the dynamic profiling. It gives an overview of the mechanics of unrolling the loop, but I thought we had just gotten through such a discussion of how this works. This description seems to start from scratch and describe it again in a different way. Maybe I'm missing something about why this is here?